src/System.IO.Pipelines.Networking.Tls/OpenSslConnectionContext.cs

```diff
@@ -0,0 +1,171 @@
+                    unencryptedData.First.TryGetPointer(out ptr);
+                    var bytesRead = Interop.SSL_write(_ssl, ptr, unencryptedData.First.Length);
+                    unencryptedData = unencryptedData.Slice(bytesRead);
+                }
+                return encryptedPipeline.Alloc().FlushAsync();
+            }
+            finally
+            {
+                CustomBio.NumberOfWrittenBytes(_writeBio);
+                handle.Free();
+            }
+        }
+
+        public Task ProcessContextMessageAsync(IPipelineWriter writer)
+        {
+            return ProcessContextMessageAsync(default(ReadableBuffer), writer);
+        }
+
+        public unsafe Task ProcessContextMessageAsync(ReadableBuffer readBuffer, IPipelineWriter writer)
+        {
+            var writeHandle = GCHandle.Alloc(writer);
+            try
+            {
+                CustomBio.SetReadBufferPointer(_readBio, ref readBuffer);
+                CustomBio.SetWriteBufferPointer(_writeBio, writeHandle);
+
+                var result = Interop.SSL_do_handshake(_ssl);
+                if (result == 1)
+                {
+                    //handshake is complete, do a final write out of data and mark as done
+                    //WriteToPipeline(ref writeBuffer, _writeBio);
+                    if (_securityContext.AplnBufferLength > 0)
+                    {
+                        byte* protoPointer;
+                        int len;
+                        Interop.SSL_get0_alpn_selected(_ssl, out protoPointer, out len);
+                        _negotiatedProtocol = ApplicationProtocols.GetNegotiatedProtocol(protoPointer, (byte)len);
+                    }
+                    _readyToSend = true;
+                    if (CustomBio.NumberOfWrittenBytes(_writeBio) > 0)
+                    {
+                        return writer.Alloc().FlushAsync();
+                    }
+                    else
+                    {
+                        return CachedTask;
+                    }
+                }
+                //We didn't get an "okay" message so lets check to see what the actual error was
+                var errorCode = Interop.SSL_get_error(_ssl, result);
+                if (errorCode == Interop.SslErrorCodes.SSL_NOTHING || errorCode == Interop.SslErrorCodes.SSL_WRITING ||
+                    errorCode == Interop.SslErrorCodes.SSL_READING)
+                {
+                    if (CustomBio.NumberOfWrittenBytes(_writeBio) > 0)
+                    {
+                        return writer.Alloc().FlushAsync();
+                    }
+                    else
+                    {
+                        return CachedTask;
+                    }
+                }
+                throw new InvalidOperationException(
+                    $"There was an error during the handshake, error code was {errorCode}");
+            }
+            catch
+            {
+                CustomBio.NumberOfWrittenBytes(_writeBio);
```

Comment:
Side effect is super important. I am carrying on the tradition... Openssl has this wonderful undocumented behavior. Even if you write a custom memory bio, if it has the "number of bytes written" > 0 when you free the bio it tries to free the memory at the pointer you stashed. Causing a process halt, very unfriendly and took 3 hours of debugging to figure out as it doesn't happen until much later (actually when you free the SslContext so no real clue what is happening). It is the joys of working with OpenSsl. Comment added.